Fix PostgreSQL JSON field search syntax error (fixes #17237)

maytham553 opened the pull request:
## Description
This PR fixes a bug with searching JSON fields in PostgreSQL. Before, search did not work and gave an error. Now, it works again.

**Fixes #17237**

- The problem started in version 3.3.31.
- The fix checks if the column already has `->>` and does not change it again.
- No changes for MySQL, SQLite, or SQL Server.

## Visual changes
No visual changes

## Functional changes
- [x] Code style has been fixed by running the `composer cs` command.
- [x] Changes have been tested to not break existing functionality.
- [x] Documentation is up-to-date.

maytham553 committed:
Fix PostgreSQL JSON field search syntax error

- Add check for existing ->> operator before processing JSON paths
- Prevents malformed SQL generation that caused syntax errors
- Fixes regression introduced in v3.3.31
- Resolves issue #17237

The fix ensures that PostgreSQL JSON field searches generate valid SQL
by properly handling cases where the ->> operator is already present
in the column expression.

maytham553 committed:
Apply code style fixes to PostgreSQL JSON search fix

maytham553 committed:
Add test cases for PostgreSQL columns with existing ->> operator

- Test basic columns with ->> operator (e.g., name->>'ar')
- Test table-prefixed columns with ->> operator (e.g., products.name->>'en')
- Test relationship columns with ->> operator (e.g., categories.name->>'ar')
- Test nested relationship columns with ->> operator (e.g., categories.parent.name->>'en')

These tests ensure our fix properly handles cases where the ->> operator
is already present in the column expression, preventing double-processing
and malformed SQL generation.

danharrin reviewed (changes requested):
Thanks.

This just skips most of the other transformation rules if `->>` already exists.

Would it not be better to preserve the behaviour of wrapping `table.field` in double quotes etc, even when `->>` is used?

Maybe your check for `->>` could be moved inside the `->map()` that is now on line 180?

maytham553 commented:
Hi @danharrin,
Just to clarify — are you suggesting doing something like this within the `map`?

```php
->map(function ($segment, $index) use ($column): string {
    // If the segment already starts with '>', preserve it (used for ->> operator)
    if (str($segment)->startsWith('>')) {
        return '->' . $segment;
    }

    $totalParts = substr_count($column, '->');

    return ($index === ($totalParts - 1))
        ? "->>'{$segment}'"
        : "->'{$segment}'";
})
```

Let me know if I understood your idea correctly or if you had something else in mind.

danharrin commented:
@maytham553 thats the right place I think, but do you not need to check if it contains `->>` already or something?

maytham553 commented:
@danharrin,

No — when `->>` is present, splitting by `'->'` removes the first `->`, so the segment we check only starts with a single `>`.
Example:

```php
$column = 'data->>name';

explode('->', $column);
// [
//   "data",
//   ">name" // starts with ">", so we detect it here
// ]
```

That’s why this already works:

```php
// If the segment starts with '>', it came from a '->>' operator.
// The first '->' was removed by explode(), so we re-add it.
if (str($segment)->startsWith('>')) {
    return '->' . $segment; // restores ->>
}
```

I’m not sure if there are any other edge cases I might be missing — what idea did you have in mind?

danharrin commented:
Okay that sounds good then, can you make those changes to the PR please?

maytham553 committed:
Refactor PostgreSQL JSON search column expression handling

- Simplified logic for processing columns with existing ->> operator.
- Updated tests to reflect changes in column expression generation, ensuring proper handling of explicit ->> operator and nested JSON paths.
- Improved clarity in test descriptions for better understanding of functionality.

danharrin commented on packages/support/src/helpers.php:
```diff
@@ -166,26 +166,38 @@ function generate_search_column_expression(string $column, ?bool $isSearchForced
+                                    ->implode('.')
+                            ) . collect(str($column)->after('->')->explode('->')) // Handle JSON path parts
+                                ->map(function ($segment, $index) use ($column): string {
+                                    // if segment already contains '>something' (from ->> operator), preserve it
```
```suggestion
                                    // If segment already starts with `>` (from `->>` operator), preserve it
```

danharrin committed:
Update packages/support/src/helpers.php

danharrin commented:
Thanks @maytham553! Considering this is well tested I will merge it in, but Glenn can still provide feedback if he wants once he has a chance.

maytham553 commented:
Thanks! — happy to adjust if @glennjacobs  spots anything.
